ulib/FStar.Interactive.PostProcess.fst

```diff
@@ -0,0 +1,760 @@
+  (* Compute the post-condition *)
+  let ge1, _, post_prop =
+    pre_post_to_propositions dbg ge0 einfo.ei_type ret_value (Some (mk_binder let_bv))
+                             tinfo einfo.ei_pre einfo.ei_post parents children
+  in
+  print_dbg dbg ("Computed post: " ^ option_to_string term_to_string post_prop);
+  (* Look for an equality in the post *)
+  let res =
+    match post_prop with
+    | None -> None
+    | Some p -> find_subequality dbg tm p
+  in
+  (* If we found something, we return the updated environment,
+   * otherwise we can return the original one *)
+  match res with
+  | None -> ge0, None
+  | Some tm -> ge1, Some tm
+
+/// Given a list of parent terms (as generated by ``explore_term``), look for an
+/// equality given by a post-condition which can be used to replace a term.
+val find_context_equality :
+     dbg:bool
+  -> ge0:genv
+  -> tm:term
+  -> parents:list term_view
+  -> children:list term_view
+  -> Tac (genv & option term)
+
+/// Auxiliary function which actually performs the search
+let rec find_context_equality_aux dbg ge0 tm (opt_bv : option bv)
+                                  (parents children : list term_view) :
+  Tac (genv & option term) =
+  match parents with
+  | [] -> ge0, None
+  | tv :: parents' ->
+    print_dbg dbg ("[> find_context_equality:\n" ^
+                   "- term  : " ^ term_to_string tm ^ "\n" ^
+                   "- parent: " ^ term_to_string tv);
+    (* We only consider let-bindings *)
+    match tv with
+    | Tv_Let _ _ bv' def _ ->
+      print_dbg dbg "Is Tv_Let";
+      let tm_info = compute_eterm_info dbg ge0.env def in
+      let einfo = tm_info.einfo in
+      (* If the searched term is a bv and the current let is the one which
+       * introduces it:
+       * - if the term is effectful, use it
+       * - otherwise, try to use its postcondition. If we don't find any
+       *   equalities, some there *)
+      let let_bv_is_tm =
+        match opt_bv with
+        | Some tm_bv -> bv_eq tm_bv bv'
+        | None -> false
+      in
+      if let_bv_is_tm && effect_type_is_pure einfo.ei_type then ge0, Some def
+      else
+        let ret_value = pack (Tv_Var bv') in
+        let ret_typ = type_of_bv bv' in
+        begin match find_equality_from_post dbg ge0 tm bv' ret_value ret_typ
+                                            einfo parents children with
+        | ge1, Some p -> ge1, Some p
+        | _, None -> find_context_equality_aux dbg ge0 tm opt_bv parents' (tv :: children)
+        end
+   | _ -> find_context_equality_aux dbg ge0 tm opt_bv parents' (tv :: children)
+
+let find_context_equality dbg ge0 tm parents children =
+  let opt_bv =
+    match inspect tm with
+    | Tv_Var bv -> Some bv
+    | _ -> None
+  in
+  find_context_equality_aux dbg ge0 tm opt_bv parents children
+
+/// Replace a subterm by another term
+val replace_term_in : bool -> term -> term -> term -> Tac term
+let rec replace_term_in dbg from_term to_term tm =
```

Comment:
Thanks for the suggestion. For this specific situation it might loop (this function actually implements general term rewriting),  however for variables substitution it's definitely worth having a look to remove those awkward (and expensive) calls to the normalizer.
Is there a way to make `visit_tm` bottom-up?

Reply:
> Is there a way to make visit_tm bottom-up?

Not currently but I could make that happen. Should we do that after this PR?

Reply:
Sure, there is no hurry: `replace_term` works well and is not a bottle-neck (contrary to variable substitution)